### PR TITLE
Refactor: extra function

### PR DESCRIPTION
**Pull request by ngaut**
Function Send is too long. This pr make it shorter.

### CODE REVIEW

**Comment by BramGruneir** [comment]
This seems like a little bit of premature optimization.  Is the doSend going to be called from elsewhere?  Or is there a logical break that's needed, say a bunch of different types of sends were called?

File: rpc/send.go [hunk 1]
```diff
@@ -128,6 +129,12 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
 			clients = append(clients, unhealthy[idx])
 		}
 	}
+
+	return doSend(clients, opts, method, addrs, getArgs, getReply)
+}
+
+func doSend(clients []*Client, opts Options, method string, addrs []net.Addr, getArgs func(addr net.Addr) interface{},
```
Comment (BramGruneir) [inline rpc/send.go]: This requires a comment describing it.

Comment (mrtracy) [inline rpc/send.go]: In the naming conventions we've been using, a method like this should be named "sendInternal" instead of "doSend".

**Comment by mrtracy** [comment]
A function being "long" does not always imply that it is too complex.  Off the top of my head, I can think of three good criteria for when a function should be reduced into subroutines:

1. If parts of the function are going to be re-used elsewhere.
2. If breaking down the function into subroutines will allow it to be unit tested more thoroughly.
3. If breaking down the function into subroutines significantly improves the clarity of the code.

This change doesn't currently fit criteria 1, this code is only used here at the moment. There is no reason to anticipate code reuse in a package internal function like this, it can be refactored on demand.

This change *may* fit criteria 2, with some additional work.  If you can create a unit test which explores `doSend()` more thoroughly by separating it from client selection, that would definitely justify this refactoring.

This doesn't really fit criteria 3 - this change is simply chaining function calls and does not improve clarity.  To improve clarity, you need to factor out routines from the *middle* of a function.  For example, If you were to also factor out the client selection routine into a `selectClients()` subroutine, that would be a better example of this criteria. The top level Send would be only a few lines long and the steps involved (`selectClients()` and `doSend()`) would be very clear. 

However, the tricky part of criteria 3 is that *short* routines like this are generally more readable inline, unless the level of indentation is excessive (it isn't here). My rule of thumb would be to leave an internal routine inline if it is less than 40 lines, otherwise consider moving it to a subroutine.  In the case of the client selection routine, it's a mere 24 lines and is easier to understand inline (with a comment) than as a subroutine.

So, in conclusion, I think you could justify this refactoring by creating a unit test that usefully tests `doSend()` (or as it should be named, `sendInternal()`) independently of the `Send()` method.  Otherwise, it would be better to leave it as a single function.

**Comment by ngaut** [comment]
@mrtracy Thanks.
I've just do some more refactoring to make the code more clear by split Send function to filterClients and sendClients. It's easy to do unit test now, Please take another look.
I haven't see your comments before the second pr. If you do doubts about the founction name, please ping me :)

**Comment by spencerkimball** [comment]
ngaut, thanks for this addition but it does seem like overkill at this stage. Feel free to improve the documentation around/inside this function instead if it was unclear to you originally.
